src/raw/client.rs

```diff
@@ -755,57 +759,45 @@ impl<PdC: PdClient> Client<PdC> {
                 max_limit: MAX_RAW_KV_SCAN_LIMIT,
             });
         }
-
-        let mut cur_range = range.into().encode_keyspace(self.keyspace, KeyMode::Raw);
+        let backoff = DEFAULT_STORE_BACKOFF;
+        let range = range.into().encode_keyspace(self.keyspace, KeyMode::Raw);
         let mut result = Vec::new();
-        let mut scan_regions = self.rpc.clone().stores_for_range(cur_range.clone()).boxed();
-        let mut region_store =
-            scan_regions
-                .next()
-                .await
-                .ok_or(Error::RegionForRangeNotFound {
-                    range: (cur_range.clone()),
-                })??;
-        let mut cur_limit = limit;
-
-        while cur_limit > 0 {
-            let request = new_raw_scan_request(
-                cur_range.clone(),
-                cur_limit,
+        let mut current_limit = limit;
+        let (start_key, end_key) = range.clone().into_keys();
+        let mut current_key: Key = start_key;
+
+        let region_error_handler =
+            |pd_rpc_client: Arc<PdC>, err: errorpb::Error, store: RegionStore| {
+                Box::pin(plan::handle_region_error(pd_rpc_client, err, store))
+            } as _;
+        while current_limit > 0 {
+            let scan_args = ScanInnerArgs {
+                start_key: current_key.clone(),
+                range: range.clone(),
```

Comment:
I was trying to trace through the logic and from what I understand, we will only be looping if the limit of the scan has not been exhausted. So regardless of a split or merge, won't we be resuming the next scan from the end_key returned by the previous scan call? So if the first scan result returns an end_key "foo", doesn't the system guarantee that if we start the next scan starting from "foo", we are guaranteed to return all results that are lexicographically larger than "foo" and smaller than whatever end_key has been provided. This is regardless of whether the underlying regions are undergoing any churn(due to any splits or merges). There may be gaps in my understanding so will be more than happy to get some more feedback here.

Reply:
> I was trying to trace through the logic and from what I understand, we will only be looping if the limit of the scan has not been exhausted. So regardless of a split or merge, won't we be resuming the next scan from the end_key returned by the previous scan call? So if the first scan result returns an end_key "foo", doesn't the system guarantee that if we start the next scan starting from "foo", we are guaranteed to return all results that are lexicographically larger than "foo" and smaller than whatever end_key has been provided. This is regardless of whether the underlying regions are undergoing any churn(due to any splits or merges). There may be gaps in my understanding so will be more than happy to get some more feedback here.

You are right. But I think the issue is here in `retryable_scan`:

```rust
        let request = new_raw_scan_request(
            scan_args.range.clone(),
            scan_args.limit,
            scan_args.key_only,
            scan_args.reverse,
            self.cf.clone(),
        );
```

As the start key in `scan_args.range` is not the end key of previous loop.

Reply:
@limbooverlambda

Reply:
Hi @pingyu. Apologies for not getting to this comment earlier, got caught up with some other deliverables. From the way I understand it, the retryable scan may not need to change the ranges since it is simply retrying the same ranges after the cache has been invalidated. The caller of retryable_scan should be changing the ranges as new regions are encountered. Let me know if that makes sense. I have however tried to simplify the logic using the loop and gotten rid of the pinned future.

Reply:
`retryable_scan` should scan from `current_key`.

Assume that we scan from `key000` to `key100`. And there are two regions: region1: `key000` ~ `key050`, region2: `key050` ~ `key100`.

First iteration, we scan kvs from region1.
Second iteration, `current_key = key050`, but this time we get an error of `EpochNotMatch` because region1 is just merged to region2, and now region2 is `key000` ~ `key100`.
Then we retry, and we should scan from `current_key`, which is `key050`, instead of `key000`. Otherwise, we will get duplicated kvs of frist iteration.

Reply:
If you look at the logic, I am. scanning from current_key in the retry instead of the start_key. Let me break down the logic:

The very first call is from the start key:

`let mut current_key: Key = start_key;` //line 166

Then we try scanning from this start key to scan:

```
let scan_args = ScanInnerArgs {
                start_key: **current_key.clone(),** // This is mutated for every iteration
                range: range.clone(),
                limit: current_limit,
                key_only,
                reverse,
                backoff: backoff.clone(),
            };
let (res, next_key) = self.retryable_scan(scan_args).await?; //This is retried with the **current state of scan_args**

```
In the retryable_scan method we are encapsulating the calls to the scan for the start_key (which is the **current_key**), now in this call if we encounter the EpochNotMatch, we will be retrying from **current_key**, not from the start_key. The retry is taking place within the function. So, no duplicate will be returned in this scenario. 

If you scroll further down in `scan_inner`, we change the current_key with the end_key returned by the `retryable_scan` function.

```
 if end_key.clone().is_some_and(|ek| ek <= next_key) {
                break;
            } else {
              current_key = next_key;
   }
```
I understand your concern but given the logic, I am having a hard time seeing how duplicates will be returned. Maybe a good way to address your concern will be to write a test that injects a fault and then change the region mappings in the mock pd and see if the test passes.

Reply:
But the `start_key` (or `from`) we pass to `new_raw_scan_request` through `range` argument is not the `scan_args.start_key`, but the `start_key` from the very beginning, the `range` argument of `scan_inner`.

```rust
            let request = new_raw_scan_request(
                scan_args.range.clone(),
                scan_args.limit,
                scan_args.key_only,
                scan_args.reverse,
                self.cf.clone(),
            );
```

Reply:
Thanks @pingyu. Now I see where the issue is. Made the change.